File: forge-prod/src/resolvers/personal-data.ts
// Personal Data Reporting — weekly scheduled trigger.
//
// Marketplace compliance hook per Atlassian's user-privacy guide:
//   https://developer.atlassian.com/platform/forge/user-privacy-guidelines/
//
// Atlassian requires every Marketplace app that stores personal data
// to periodically report the accountIds it holds and act on
// Atlassian's anonymization decisions. The protocol is:
//
//   1. Collect every accountId we hold + the latest updatedAt per account.
//   2. POST batches of <= 90 accounts to
//      https://api.atlassian.com/app/report-accounts/ with
//      `report:personal-data` scope (api.asApp() handles auth).
//   3. Atlassian returns per-account `status` — "closed" means the
//      account was deleted/anonymized; "updated" means the user info
//      changed (we ignore — webhook + daily reconcile already refresh).
//   4. Erase data for "closed" accounts.
//
// Run cadence: weekly (every Sunday). The default cycle from
// Atlassian's docs is 7 days; running once a week comfortably stays
// within that cycle while keeping invocation count low.
//
// Per-tenant scoping: this resolver is invoked once per install via
// the scheduled trigger. The backend's /api/forge/personal-data/* endpoints
// are tenant-scoped via the FIT auth middleware, so no tenant_id needs
// to flow through the wire — invokeRemote attaches the FIT and the
// backend resolves the tenant from there.

import api, { invokeRemote, route } from "@forge/api";

const BACKEND_REMOTE_KEY = "backend";
const ATLASSIAN_REPORT_ENDPOINT = "https://api.atlassian.com/app/report-accounts/";

// Atlassian doc: max 90 accounts per request.
const ATLASSIAN_BATCH_SIZE = 90;

// Per-page accountIds we pull from the backend per HTTP round-trip.
// Larger pages = fewer round-trips but larger payloads. 200 is well
// within both Forge's 25s invocation cap and Atlassian's request size
// limits at the report endpoint.
const BACKEND_PAGE_SIZE = 200;

interface AccountEntry {
  account_id: string;
  updated_at: string; // RFC 3339
}

interface AccountsResponse {
  accounts: AccountEntry[];
  next_cursor: string | null;
}

interface AtlassianAccountStatus {
  accountId: string;
  status: "closed" | "updated" | string;
}

interface AtlassianReportResponse {
  accounts?: AtlassianAccountStatus[];
}

async function fetchAllAccounts(): Promise<AccountEntry[]> {
  const all: AccountEntry[] = [];
  let cursor: string | null = null;
  // Hard guard against an unbounded loop if the backend ever
  // mis-implements paging — abort after 100 pages (= 20k accounts at
  // current page size, well above any plausible single-tenant volume).
  for (let i = 0; i < 100; i++) {
    const path = cursor
      ? `/api/forge/personal-data/accounts?limit=${BACKEND_PAGE_SIZE}&cursor=${encodeURIComponent(cursor)}`
      : `/api/forge/personal-data/accounts?limit=${BACKEND_PAGE_SIZE}`;
    const res = await invokeRemote(BACKEND_REMOTE_KEY, {
      path,
      method: "GET",
      headers: { "Content-Type": "application/json" },
    });
    if (!res.ok) {
      throw new Error(`Backend accounts fetch failed: ${res.status}`);
    }
    const body = (await res.json()) as AccountsResponse;
    all.push(...body.accounts);
    if (!body.next_cursor) {
      return all;
    }
    cursor = body.next_cursor;
  }
  throw new Error("Backend accounts pagination did not terminate after 100 pages");
}

function chunk<T>(items: T[], size: number): T[][] {
  const out: T[][] = [];
  for (let i = 0; i < items.length; i += size) {
    out.push(items.slice(i, i + size));
  }
  return out;
}

async function reportBatch(batch: AccountEntry[]): Promise<string[]> {
  // Returns the accountIds Atlassian flagged as "closed" — these need
  // to be erased on our side. "updated" entries are intentionally
  // ignored: webhooks + daily reconcile already refresh display names.
  const res = await api.fetch(ATLASSIAN_REPORT_ENDPOINT, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      // Atlassian's expected shape:
      // { accounts: [{ accountId, updatedAt }, ...] }
      accounts: batch.map((a) => ({
        accountId: a.account_id,
        updatedAt: a.updated_at,
      })),
    }),
  });
  if (!res.ok) {
    throw new Error(`Atlassian report-accounts failed: ${res.status} ${await res.text()}`);
  }
  const body = (await res.json()) as AtlassianReportResponse;
  return (body.accounts ?? [])
    .filter((entry) => entry.status === "closed")
    .map((entry) => entry.accountId);
}

async function eraseAccounts(accountIds: string[]): Promise<void> {
  if (accountIds.length === 0) return;
  const res = await invokeRemote(BACKEND_REMOTE_KEY, {
    path: "/api/forge/personal-data/erase",
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ account_ids: accountIds }),
  });
  if (!res.ok) {
    throw new Error(`Backend erase failed: ${res.status} ${await res.text()}`);
  }
  const body = (await res.json()) as { issues_updated: number };
  console.log(
    `Erased ${accountIds.length} accountId(s); ${body.issues_updated} issue rows updated.`,
  );
}

export const personalDataReportingResolver = async () => {
  const startedAt = Date.now();
  try {
    const accounts = await fetchAllAccounts();
    if (accounts.length === 0) {
      console.log("Personal data reporting: tenant has no accountIds; skipping.");
      return;
    }

    const batches = chunk(accounts, ATLASSIAN_BATCH_SIZE);
    const closedAccountIds: string[] = [];
    for (const batch of batches) {
      const closed = await reportBatch(batch);
      closedAccountIds.push(...closed);
    }

    await eraseAccounts(closedAccountIds);
    console.log(
      `Personal data reporting cycle complete: reported ${accounts.length} accountId(s) in ${batches.length} batch(es); erased ${closedAccountIds.length}; ${Date.now() - startedAt}ms.`,
    );
  } catch (e) {
    // Don't swallow; let the Forge runtime log + retry on next cycle.
    // Atlassian doesn't penalize a single missed cycle, but consecutive
    // misses risk the listing flagging non-compliance.
    console.error(`Personal data reporting cycle failed: ${(e as Error).message}`);
    throw e;
  }
};

// `route` is exported in this module purely to keep the import lint-clean
// for future endpoints that need typed Jira routes from this file.
export { route };

File: backend/app/routers/personal_data.py
"""Personal Data Reporting endpoints — Marketplace compliance hook.

Atlassian requires every Marketplace app that stores personal data to
periodically report the accountIds it holds and act on Atlassian's
anonymization decisions. The protocol is documented at
https://developer.atlassian.com/platform/forge/user-privacy-guidelines/.

This router exposes two tenant-scoped endpoints used by the Forge
weekly scheduled trigger (`personalDataReportingResolver` in
forge-prod/src/resolvers/personal-data.ts):

- `GET  /api/forge/personal-data/accounts?cursor=<id>&limit=<n>`
  Returns up to `limit` (default 200) distinct accountIds the tenant
  has stored, with an `updatedAt` timestamp per account (the most
  recent issue.updated_at across that account's issues). Cursor-based
  paging on accountId so the resolver can chunk through tenants with
  many users without hitting Forge's 25 sec invocation limit.

- `POST /api/forge/personal-data/erase`
  Body: `{"account_ids": ["abc", "def", ...]}`. For each accountId,
  null out `assignee` + `assignee_account_id` on every issue row in
  this tenant. Used after the resolver POSTs to Atlassian's
  `report-accounts` API and receives `closed` statuses.

Both endpoints require Forge tenant context — they go through the
standard auth middleware and use `current_tenant_context` to scope
queries.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field
from sqlalchemy import select, update
from sqlalchemy.orm import Session

from app.core.deps import current_tenant_context
from app.core.logging import get_logger
from app.core.tenant_context import TenantContext
from app.db.models import Issue
from app.db.session import get_db

logger = get_logger(__name__)

router = APIRouter(prefix="/forge/personal-data", tags=["forge-personal-data"])


class AccountEntry(BaseModel):
    account_id: str
    updated_at: datetime


class AccountsResponse(BaseModel):
    accounts: list[AccountEntry]
    next_cursor: str | None = Field(
        default=None,
        description=(
            "Pass back as `?cursor=<id>` to fetch the next page. "
            "`null` when the caller has reached the end."
        ),
    )


class EraseRequest(BaseModel):
    account_ids: list[str] = Field(default_factory=list, max_length=500)


class EraseResponse(BaseModel):
    erased_account_ids: list[str]
    issues_updated: int


@router.get("/accounts", response_model=AccountsResponse)
def list_accounts(
    cursor: str | None = Query(
        default=None,
        description="accountId from the previous page's `next_cursor`. Omit on first call.",
    ),
    limit: int = Query(default=200, ge=1, le=500),
    db: Session = Depends(get_db),
    ctx: TenantContext = Depends(current_tenant_context),
) -> AccountsResponse:
    """Return distinct accountIds with their most recent issue.updated_at.

    The resolver paginates through this until `next_cursor` is null,
    then batches the accumulated accountIds (max 90 per Atlassian doc)
    to `https://api.atlassian.com/app/report-accounts/`.
    """
    # Distinct accountIds with the max(updated_at) per id, ordered by
    # accountId so the cursor is monotonic (cursor = "give me ids
    # strictly greater than this"). NULLs (unassigned issues) are
    # excluded — there's no personal data to report for them.
    from sqlalchemy import func

    base = (
        select(
            Issue.assignee_account_id,
            func.max(Issue.updated_at).label("updated_at"),
        )
        .where(
            Issue.tenant_id == ctx.tenant_id,
            Issue.assignee_account_id.is_not(None),
        )
        .group_by(Issue.assignee_account_id)
        .order_by(Issue.assignee_account_id)
        .limit(limit + 1)
    )
    if cursor is not None:
        base = base.where(Issue.assignee_account_id > cursor)

    rows: list[Any] = list(db.execute(base).all())
    has_more = len(rows) > limit
    page = rows[:limit]

    next_cursor: str | None = None
    if has_more and page:
        next_cursor = page[-1][0]

    return AccountsResponse(
        accounts=[AccountEntry(account_id=r[0], updated_at=r[1]) for r in page],
        next_cursor=next_cursor,
    )


@router.post("/erase", response_model=EraseResponse)
def erase_accounts(
    body: EraseRequest,
    db: Session = Depends(get_db),
    ctx: TenantContext = Depends(current_tenant_context),
) -> EraseResponse:
    """Null out assignee + assignee_account_id for the given accounts.

    Idempotent: re-running with the same accountIds is a no-op once the
    first run has erased the data. Tenant-scoped via ctx.tenant_id, so
    one tenant cannot trigger erasure for another tenant's data even
    if accountIds overlapped.
    """
    if not body.account_ids:
        return EraseResponse(erased_account_ids=[], issues_updated=0)

    stmt = (
        update(Issue)
        .where(
            Issue.tenant_id == ctx.tenant_id,
            Issue.assignee_account_id.in_(body.account_ids),
        )
        .values(assignee=None, assignee_account_id=None)
    )
    result = db.execute(stmt)
    db.commit()
    issues_updated = result.rowcount or 0

    logger.info(
        "Erased personal data for tenant=%s accounts=%d issues=%d",
        ctx.tenant_id,
        len(body.account_ids),
        issues_updated,
    )
    return EraseResponse(
        erased_account_ids=body.account_ids,
        issues_updated=issues_updated,
    )
